testutil/apiload/main.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	// If CPUs are specified, set a max
+	if Config.CPUs >= 1 {
+		runtime.GOMAXPROCS(Config.CPUs)
+		log.Printf("Running on %d CPUs.", Config.CPUs)
+	}
+
+	var wg sync.WaitGroup
+	resChan := make(chan *vegeta.Result)
+
+	// Duration from profile can be overridden in config or cli argument.
+	seconds := attackProfile.Duration
+	if Config.Duration > 0 {
+		seconds = Config.Duration
+	}
+
+	// Perform the load test.
+	log.Printf("Beginning %s. Duration: %d seconds.", Config.Attack, seconds)
+	duration := time.Duration(seconds*1000) * time.Millisecond
+	for idx, attacker := range attackProfile.Attackers {
+		if len(attacker.Endpoints) == 0 {
+			log.Fatalf("Empty enpoint list encountered for %s", Config.Attack)
+		}
+		rate := vegeta.Rate{
+			Freq: attacker.Frequency,
+			Per:  time.Second,
+		}
+
+		var targets []vegeta.Target
+		for _, endpoint := range attacker.Endpoints {
+			path := Config.Server + endpoint
+			targets = append(targets, vegeta.Target{
+				Method: "GET",
+				URL:    path,
+			})
+		}
+
+		targeter := vegeta.NewStaticTargeter(targets...)
+		// Truncate all response bodies at 100 bytes
+		vegAttacker := vegeta.NewAttacker(vegeta.MaxBody(100))
+
+		wg.Add(1)
+		go func(attacker *Attacker) {
+			defer wg.Done()
+			remainder := duration
+			if attacker.Delay > 0 {
+				if attacker.Delay >= float32(attackProfile.Duration) {
+					log.Printf("The specified delay, %f, for attacker %d is >= the duration of test, %d.", attacker.Delay, idx, Config.Attack)
+					return
+				}
+				attackDelay := time.Duration(int(attacker.Delay*1000)) * time.Millisecond
+				time.Sleep(attackDelay)
+				remainder -= attackDelay
+			}
+			if attacker.Name != "" {
+				log.Printf("Beggining %s", attacker.Name)
+			}
+			for res := range vegAttacker.Attack(targeter, rate, remainder, fmt.Sprintf("%s:%d", Config.Attack, idx)) {
+				resChan <- res
+			}
+		}(attacker)
+	}
+
+	// Make a notification channel for the end of the attack.
+	c := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(c)
+	}()
+
+	// A couple of helper functions.
+	// Vegeta doesn't return the url in the result, so look it up,
+	// finding the Attacker along the way.
+	findAttacker := func(result *vegeta.Result) (attacker *Attacker, endpoint string) {
+		parts := strings.Split(result.Attack, ":")
+		attackerIdx, err := strconv.Atoi(parts[len(parts)-1])
+		if err != nil {
+			log.Fatalf("Failed to parse attacker index. Something is terribly wrong.")
+			return
+		}
+		attacker = attackProfile.Attackers[attackerIdx]
+		endpoint = attacker.Endpoints[int(result.Seq)%len(attacker.Endpoints)]
+		return
+	}
+
+	// A list of unique error URLs. Print messages for new ones. Ignore repeats.
+	var problematic []string
+	addProblematic := func(result *vegeta.Result) {
+		_, endpoint := findAttacker(result)
+		for _, url := range problematic {
+			if url == endpoint {
+				return
+			}
+		}
+		problematic = append(problematic, endpoint)
+		log.Printf("Error code %d from %s: %s",
```

Comment:
When code is 0, the connection failed, but this is not necessarily obvious.  I think it is worth special casing a log message for code 0.